examples/all-clusters-app/all-clusters-common/src/EnergyEvseDelegateImpl.cpp

```diff
@@ -0,0 +1,771 @@
+ */
+Status EnergyEvseDelegate::HwSetVehicleID(const CharSpan & newValue)
+{
+    if (mVehicleID.IsNull() || !newValue.data_equal(mVehicleID.Value()))
+    {
+        /* create a copy of the string so the callee doesn't have to keep it */
+        char * destinationBuffer = new char[kMaxVehicleIDBufSize];
+
+        MutableCharSpan destinationString(destinationBuffer, kMaxVehicleIDBufSize);
+        CHIP_ERROR err = CopyCharSpanToMutableCharSpan(newValue, destinationString);
+        if (err != CHIP_NO_ERROR)
+        {
+            ChipLogError(AppServer, "HwSetVehicleID - could not copy vehicleID");
+            delete[] destinationBuffer;
+            return Status::Failure;
+        }
+
+        if (!mVehicleID.IsNull())
+        {
+            delete[] mVehicleID.Value().data();
+        }
+
+        mVehicleID = MakeNullable(static_cast<CharSpan>(destinationString));
+
+        ChipLogDetail(AppServer, "VehicleID updated");
+        MatterReportingAttributeChangeCallback(mEndpointId, EnergyEvse::Id, VehicleID::Id);
+    }
+
+    return Status::Success;
+}
+
+/* ---------------------------------------------------------------------------
+ * Functions below are private helper functions internal to the delegate
+ */
+
+/**
+ *  @brief   Called to compute the safe charging current limit
+ */
+Status EnergyEvseDelegate::ComputeMaxChargeCurrentLimit()
+{
+    int64_t oldValue;
+    /* mActualChargingCurrentLimit is the minimum of:
+     *   - MaxHardwareCurrentLimit (of the hardware)
+     *   - CircuitCapacity (set by the electrician - less than the hardware)
+     *   - CableAssemblyLimit (detected when the cable is inserted)
+     *   - MaximumChargeCurrent (from charging command)
+     *   - UserMaximumChargeCurrent (could dynamically change)
+     */
+
+    oldValue                    = mActualChargingCurrentLimit;
+    mActualChargingCurrentLimit = mMaxHardwareCurrentLimit;
+    mActualChargingCurrentLimit = min(mActualChargingCurrentLimit, mCircuitCapacity);
+    mActualChargingCurrentLimit = min(mActualChargingCurrentLimit, mCableAssemblyCurrentLimit);
+    mActualChargingCurrentLimit = min(mActualChargingCurrentLimit, mMaximumChargingCurrentLimitFromCommand);
+    mActualChargingCurrentLimit = min(mActualChargingCurrentLimit, mUserMaximumChargeCurrent);
+
+    /* Set the actual max charging current attribute */
+    mMaximumChargeCurrent = mActualChargingCurrentLimit;
+
+    if (oldValue != mMaximumChargeCurrent)
+    {
+        ChipLogDetail(AppServer, "MaximumChargeCurrent updated to %ld", static_cast<long>(mMaximumChargeCurrent));
+        MatterReportingAttributeChangeCallback(mEndpointId, EnergyEvse::Id, MaximumChargeCurrent::Id);
+
+        /* Call the EV Charger hardware current limit callback */
+        // TODO
+    }
+    return Status::Success;
+}
+
+/**
+ * Attribute methods
+ */
+/* State */
+StateEnum EnergyEvseDelegate::GetState()
+{
+    return mState;
+}
+CHIP_ERROR EnergyEvseDelegate::SetState(StateEnum newValue)
```

Comment:
supernit: blank space above to separate methods. Maybe clang-format should reformat this file ... I am surprised it did not add it.

Reply:
more instances below.

Reply:
Fixed.